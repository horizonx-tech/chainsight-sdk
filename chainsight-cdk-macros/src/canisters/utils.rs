use std::path::PathBuf;

pub fn extract_contract_name_from_path(s: &str) -> String {
    let path = PathBuf::from(s);
    let name = path.file_stem().expect("file_stem failed");
    name.to_str().expect("to_str failed").to_string()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_extract_contract_name_from_path() {
        let path = "__interfaces/Oracle.json";
        assert_eq!(extract_contract_name_from_path(path), "Oracle");
    }
}
