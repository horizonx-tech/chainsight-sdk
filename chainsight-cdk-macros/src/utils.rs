use proc_macro::TokenStream;
use quote::quote;
use syn::parse_macro_input;

pub fn chainsight_common(input: TokenStream) -> TokenStream {
    let item = parse_macro_input!(input as syn::LitInt);
    let output = quote! {
        #[derive(Clone, Debug, PartialEq, candid::CandidType, candid::Deserialize, serde::Serialize)]
        pub struct CanisterMetricsSnapshot {
            pub timestamp: u64,
            pub cycles: u128,
        }
        chainsight_cdk_macros::manage_vec_state!("canister_metrics_snapshot", CanisterMetricsSnapshot, false);

        fn setup_monitoring_canister_metrics() {
            let round_timestamp = |ts: u32, unit: u32| ts / unit * unit;
            let current_time_sec = (ic_cdk::api::time() / (1000 * 1000000)) as u32;
            let delay = round_timestamp(current_time_sec, #item) + #item - current_time_sec;

            ic_cdk_timers::set_timer(std::time::Duration::from_secs(delay as u64), move || {
                ic_cdk_timers::set_timer_interval(std::time::Duration::from_secs(#item as u64), || {
                    monitor_canister_metrics();
                });
            });
            // in first time, immediate execution
            monitor_canister_metrics();
        }

        fn monitor_canister_metrics() {
            let timestamp = ic_cdk::api::time();
            let cycles = ic_cdk::api::canister_balance128();
            let datum = CanisterMetricsSnapshot {
                timestamp,
                cycles,
            };
            ic_cdk::println!("monitoring: {:?}", datum.clone());
            add_canister_metrics_snapshot(datum);
        }

        #[ic_cdk_macros::init]
        fn init() {
            setup_monitoring_canister_metrics()
        }

        #[ic_cdk_macros::post_upgrade]
        fn post_upgrade() {
            setup_monitoring_canister_metrics()
        }

        #[ic_cdk::query]
        #[candid::candid_method(query)]
        pub fn metric() -> CanisterMetricsSnapshot {
            get_last_canister_metrics_snapshot()
        }

        #[ic_cdk::query]
        #[candid::candid_method(query)]
        pub fn metrics(n: usize) -> Vec<CanisterMetricsSnapshot> {
            get_top_canister_metrics_snapshots(n)
        }

        async fn _get_target_proxy(target: candid::Principal) -> candid::Principal {
            let out: ic_cdk::api::call::CallResult<(candid::Principal,)> = ic_cdk::api::call::call(target, "get_proxy", ()).await;
            out.unwrap().0
        }
    };
    output.into()
}

pub fn did_export(input: TokenStream) -> TokenStream {
    let item = parse_macro_input!(input as syn::LitStr);
    let file_name = item.value() + ".did";
    TokenStream::from(quote! {
        candid::export_service!();
        #[ic_cdk::query(name = "__get_candid_interface_tmp_hack")]
        #[candid::candid_method(query, rename = "__get_candid_interface_tmp_hack")]
        fn __export_did_tmp_() -> String {
            __export_service()
        }
        #[cfg(test)]
        mod tests {
            use super::*;

            #[test]
            fn gen_candid() {
                std::fs::write(#file_name, __export_service()).unwrap();
            }
        }
    })
}
