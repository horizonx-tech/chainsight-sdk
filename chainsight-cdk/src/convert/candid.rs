use std::{cmp::max, fs, ops::Deref, path::Path};

use anyhow::Ok;
use candid::{
    bindings::rust::{compile, Target},
    check_prog,
    types::{Type, TypeInner},
    IDLProg, TypeEnv,
};
use regex::Regex;

#[derive(Debug, Clone)]
pub struct CanisterMethodIdentifier {
    pub identifier: String,
    type_env: TypeEnv,
}
impl CanisterMethodIdentifier {
    pub const REQUEST_ARGS_TYPE_NAME: &'static str = "RequestArgsType";
    pub const RESPONSE_TYPE_NAME: &'static str = "ResponseType";
    pub const SUFFIX_IN_DID: &'static str = "InDid";

    pub fn new(s: &str) -> anyhow::Result<Self> {
        Self::new_internal(s, None)
    }

    pub fn new_with_did(s: &str, dependended_did: String) -> anyhow::Result<Self> {
        // NOTE: If the reserved Type specified in the .did to import is used, a Compile error will occur due to duplication.
        let (s, dependended_did) = avoid_using_reserved_types(s, &dependended_did);
        Self::new_internal(&s, Some(dependended_did))
    }

    fn new_internal(s: &str, dependended_did: Option<String>) -> anyhow::Result<Self> {
        let (identifier, args_ty, response_ty) = extract_elements(s)?;
        let did: String = Self::generate_did(&args_ty, &response_ty);

        let ast: IDLProg = if let Some(base_did) = dependended_did {
            format!("{}\n\n{}", base_did, did)
        } else {
            did.to_string()
        }
        .parse()?;
        let mut type_env = TypeEnv::new();
        let _ = check_prog(&mut type_env, &ast);

        Ok(Self {
            identifier,
            type_env,
        })
    }

    pub fn compile(&self) -> anyhow::Result<String> {
        anyhow::ensure!(self.compilable(), "Not compilable IDLProg");

        let mut config = candid::bindings::rust::Config::new();
        // // Update the structure derive to chainsight's own settings
        config.type_attributes = "#[derive(Clone, Debug, candid :: CandidType, candid :: Deserialize, serde :: Serialize, chainsight_cdk_macros :: StableMemoryStorable)]".to_string();
        config.target = Target::CanisterStub;
        let contents = compile(&config, &self.type_env, &None);

        let mut lines = contents
            .lines()
            // Delete initial 'use' declarations
            .skip(4)
            // Delete comment lines and blank lines
            .filter(|line| !(line.starts_with("//") || line.is_empty()))
            .map(|line| {
                // convert icp's own Nat and Int types to rust native type
                // NOTE: num-traits can be used, but is not used to reduce dependencies
                //  https://forum.dfinity.org/t/candid-nat-to-u128/16016
                //  https://discord.com/channels/748416164832608337/872791506853978142/1162494173933481984
                // NOTE: when ready to convert to u128/i128, consider with EthAbiEncoder's Encoder trait
                if line.contains("candid::Nat") {
                    return line.replace("candid::Nat", "u128");
                }
                if line.contains("candid::Int") {
                    return line.replace("candid::Int", "i128");
                }
                line.to_string()
            })
            .collect::<Vec<_>>();
        lines.insert(0, "#![allow(dead_code, unused_imports)]".to_string());
        lines.insert(
            1,
            "use candid::{self, CandidType, Deserialize, Principal, Encode, Decode};".to_string(),
        );
        Ok(lines.join("\n"))
    }

    pub fn get_types(&self) -> (Option<&Type>, Option<&Type>) {
        (
            self.find_type(Self::REQUEST_ARGS_TYPE_NAME),
            self.find_type(Self::RESPONSE_TYPE_NAME),
        )
    }

    pub fn get_type(&self, key: &str) -> Option<&Type> {
        self.find_type(key)
    }

    fn compilable(&self) -> bool {
        let (args_ty, response_ty) = self.get_types();
        let not_compilable_type = &TypeInner::Unknown;

        if args_ty.is_some_and(|ty| ty.deref().eq(not_compilable_type)) {
            return false;
        }
        if response_ty.is_some_and(|ty| ty.deref().eq(not_compilable_type)) {
            return false;
        }

        true
    }

    fn generate_did(args_ty: &str, response_ty: &str) -> String {
        let args_ty_did = if args_ty.is_empty() {
            "".to_string()
        } else {
            generate_did_type(Self::REQUEST_ARGS_TYPE_NAME, args_ty)
        };
        let response_ty_did = if response_ty.is_empty() {
            "".to_string()
        } else {
            generate_did_type(Self::RESPONSE_TYPE_NAME, response_ty)
        };
        format!("{}\n{}", args_ty_did, response_ty_did)
    }

    fn find_type(&self, key: &str) -> Option<&Type> {
        let ty = self.type_env.find_type(key);
        ty.ok()
    }
}

fn generate_did_type(key: &str, value: &str) -> String {
    format!("type {} = {};", key, value)
}

fn extract_elements(s: &str) -> anyhow::Result<(String, String, String)> {
    let (identifier, remains) = s
        .split_once(':')
        .expect("Invalid canister method identifier");
    let (args_ty, response_ty) = remains
        .split_once("->")
        .expect("Invalid canister method identifier");

    let trim_type_str = |s: &str| {
        let trimed = s.trim();
        let removed_brackets = trimed.trim_matches(|c| c == '(' || c == ')');
        removed_brackets.trim().to_string()
    };

    Ok((
        identifier.trim().to_string(),
        trim_type_str(args_ty),
        trim_type_str(response_ty),
    ))
}

fn avoid_using_reserved_types(s: &str, did: &str) -> (String, String) {
    let req_ty = CanisterMethodIdentifier::REQUEST_ARGS_TYPE_NAME;
    let res_ty = CanisterMethodIdentifier::RESPONSE_TYPE_NAME;
    let base_suffix = CanisterMethodIdentifier::SUFFIX_IN_DID;

    // Check maximum value of suffix for reserved type
    let pattern = format!(
        r"(?P<name>({}|{}))_{}_(?P<suffix_num>\d+)",
        req_ty, res_ty, base_suffix
    );
    let re = Regex::new(&pattern).expect("Invalid regex pattern");
    let mut max_number = 0;
    for line in did.lines() {
        for cap in re.captures_iter(line) {
            let suffix_num = cap.name("suffix_num").unwrap().as_str();
            if let core::result::Result::Ok(number) = suffix_num.parse::<u32>() {
                max_number = max(max_number, number);
            }
        }
    }
    max_number += 1;

    // Replace reserved type (did)
    let pattern = format!(r" (?P<name>({}|{}))(?P<last_char>(;|,| ))", req_ty, res_ty);
    let re = Regex::new(&pattern).expect("Invalid regex pattern");
    let replace_reserved_type = |s: &str| {
        let replaced_s = re.replace_all(s, |caps: &regex::Captures| {
            let name = caps.name("name").unwrap().as_str();
            let last_char = caps.name("last_char").unwrap().as_str();
            format!(" {}_{}_{}{}", name, base_suffix, max_number, last_char)
        });
        replaced_s.to_string()
    };
    let replaced_did_lines = did.lines().map(replace_reserved_type).collect::<Vec<_>>();

    // Replace reserved type (identifier)
    let pattern = format!(
        r"(?P<first_char>(\(| ))(?P<name>({}|{}))(?P<last_char>(\)|;|,| ))",
        req_ty, res_ty
    );
    let re = Regex::new(&pattern).expect("Invalid regex pattern");
    let replaced_s = re.replace_all(s, |caps: &regex::Captures| {
        let first_char = caps.name("first_char").unwrap().as_str();
        let name = caps.name("name").unwrap().as_str();
        let last_char = caps.name("last_char").unwrap().as_str();
        format!(
            "{}{}_{}_{}{}",
            first_char, name, base_suffix, max_number, last_char
        )
    });

    (replaced_s.to_string(), replaced_did_lines.join("\n"))
}

// Read .did and remove 'service' section
pub fn read_did_to_string_without_service<P: AsRef<Path>>(path: P) -> anyhow::Result<String> {
    let mut data = fs::read_to_string(path)?;
    Ok(exclude_service_from_did_string(&mut data))
}
fn exclude_service_from_did_string(data: &mut String) -> String {
    // remove 'service' section
    if let Some(service_start_idx) = data.find("service : {") {
        data.truncate(service_start_idx);
    }

    data.trim().to_owned()
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_snapshot;

    const TEST_IDENTS: &'static [(&'static str, &'static str); 6] = &[
        (&"normal", &"update_value : (nat64) -> (text)"),
        (&"normal_nat", &"use_nat : (nat32) -> (nat)"),
        (&"normal_int", &"use_nat : (int32) -> (int)"),
        (&"no_args", &"get_value : () -> (text)"),
        (
            &"record",
            &"get_snapshot : (nat64) -> (record { value : text; timestamp : nat64 })",
        ),
        (
            &"nested_record",
            &"get_last_snapshot : () -> (record { dai : record { usd : float64 } })",
        ),
    ];

    const TEST_IDENTS_WITH_DID: &'static [(&'static str, &'static str, &'static str); 5] = &[
        (
            &"single",
            &"get_snapshot : (nat64) -> (Snapshot)",
            &"type Snapshot = record { value : text; timestamp : nat64 };",
        ),
        (
            &"multiple",
            &"get_snapshot : (nat64) -> (Snapshot)",
            &r#"type CurrencyValue = record { usd : float64 };
type Snapshot = record { value : SnapshotValue; timestamp : nat64 };
type SnapshotValue = record { dai : CurrencyValue };
"#,
        ),
        (
            &"with_variant",
            &"init_in : (Env) -> (Result)",
            &r#"type Env = variant { Production; Test; LocalDevelopment };
type InitError = variant {
  InvalidDestination : text;
  InvalidPrincipal : principal;
  InvalidContent : text;
  InvalidRequest : text;
};
type Result = variant { Ok; Err : InitError };
"#,
        ),
        (
            &"with_vec",
            &"get_sources : () -> (vec Sources)",
            &r#"type HttpsSnapshotIndexerSourceAttrs = record {
  queries : vec record { text; text };
};
type SourceType = variant { evm; https; chainsight };
type Sources = record {
  source : text;
  interval_sec : opt nat32;
  attributes : HttpsSnapshotIndexerSourceAttrs;
  source_type : SourceType;
};"#,
        ),
        (
            &"with_reserved_type",
            &"get_snapshot : (RequestArgsType) -> (ResponseType)",
            &r#"type RequestArgsType = nat64;
type ResponseType = text;"#,
        ),
    ];

    #[test]
    fn test_compile() {
        for (label, s) in TEST_IDENTS {
            let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
            let compiled = ident.compile().unwrap();
            assert_snapshot!(format!("compile__{}", label.to_string()), compiled);
        }
    }

    #[test]
    fn test_compile_with_depended_did() {
        for (label, s, did) in TEST_IDENTS_WITH_DID {
            let ident = CanisterMethodIdentifier::new_with_did(s, did.to_string())
                .expect("Failed to parse");
            let compiled = ident.compile().unwrap();
            assert_snapshot!(
                format!("compile_with_depended_did__{}", label.to_string()),
                compiled
            );
        }
    }

    #[test]
    fn test_compile_when_not_compilable() {
        let s = "get_snapshot : (Input) -> (Output)";
        let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
        assert_eq!(
            ident.compile().err().unwrap().to_string(),
            "Not compilable IDLProg"
        );
    }

    #[test]
    fn test_compilable() {
        // compilable
        for (_, s) in TEST_IDENTS {
            let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
            assert!(ident.compilable());
        }
        for (_, s, did) in TEST_IDENTS_WITH_DID {
            let ident = CanisterMethodIdentifier::new_with_did(s, did.to_string())
                .expect("Failed to parse");
            assert!(ident.compilable());
        }

        // not compilable
        let not_compilable = vec![
            "get_snapshot : (Input) -> (Output)",
            "get_snapshot : (nat64) -> (Snapshot)",
            "get_snapshot : (Snapshot) -> (nat64)",
        ];
        for s in not_compilable {
            let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
            assert!(!ident.compilable());
        }

        let not_compilable_with_did = vec![
            (
                "get_snapshot : (Snapshot) -> (Snapshot)",
                "type Snapshot_1 = record { value : text; timestamp : nat64 };",
            ),
            (
                "get_snapshot : (nat64) -> (Snapshot)",
                "type Snapshot_1 = record { value : text; timestamp : nat64 };",
            ),
            (
                "get_snapshot : (Snapshot) -> (nat64)",
                "type Snapshot_1 = record { value : text; timestamp : nat64 };",
            ),
        ];
        for (s, did) in not_compilable_with_did {
            let ident = CanisterMethodIdentifier::new_with_did(s, did.to_string())
                .expect("Failed to parse");
            assert!(!ident.compilable());
        }
    }

    #[test]
    fn test_generate_did() {
        let key = "Account";
        let value = "record { owner : principal; subaccount : opt blob }";
        let did = generate_did_type(key, value);
        assert_snapshot!(did);
    }

    #[test]
    fn test_extract_elements() {
        for (label, s) in TEST_IDENTS {
            let result = extract_elements(s).expect("Failed to parse");
            assert_snapshot!(
                format!("extract_elements__{}", label.to_string()),
                format!("{:#?}", result)
            );
        }
    }

    #[test]
    fn test_extract_elements_multi_nested_record() {
        let s =
            "icrc2_allowance : (record { account : record { owner : principal; subaccount : opt blob }; spender : record { owner : principal; subaccount : opt blob } }) -> (record { allowance : nat; expires_at : opt nat64 })";
        let (identifier, args_ty, response_ty) = extract_elements(s).expect("Failed to parse");
        assert_eq!(identifier, "icrc2_allowance");
        assert_eq!(args_ty, "record { account : record { owner : principal; subaccount : opt blob }; spender : record { owner : principal; subaccount : opt blob } }");
        assert_eq!(
            response_ty,
            "record { allowance : nat; expires_at : opt nat64 }"
        );
    }

    #[test]
    fn test_exclude_service_from_did_string() {
        let expected = r#"type List = opt record { head: int; tail: List };
type byte = nat8;".to_string()"#;
        let mut actual = format!("{}\n", expected)
            + r#"service : {
  f : (byte, int, nat, int8) -> (List);
  g : (List) -> (int) query;
}"#;

        assert_eq!(exclude_service_from_did_string(&mut actual), expected);
    }
}
