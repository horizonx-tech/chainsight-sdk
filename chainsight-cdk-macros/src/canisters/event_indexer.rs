use std::{fs::File, path::Path};

use chainsight_cdk::{
    config::components::{CommonConfig, EventIndexerConfig, EventIndexerEventDefinition},
    convert::evm::{convert_type_from_ethabi_param_type, ADDRESS_TYPE, U256_TYPE},
};
use ethabi;
use proc_macro::TokenStream;
use quote::{format_ident, quote};
use syn::parse_macro_input;

use crate::canisters::utils::camel_to_snake;

use super::utils::extract_contract_name_from_path;

pub fn def_event_indexer_canister(input: TokenStream) -> TokenStream {
    let input_json_string: String = parse_macro_input!(input as syn::LitStr).value();
    let config: EventIndexerConfig =
        serde_json::from_str(&input_json_string).expect("Failed to parse input_json_string");
    event_indexer_canister(config).into()
}

fn event_indexer_canister(config: EventIndexerConfig) -> proc_macro2::TokenStream {
    let common = common_code(&config.common);
    let custom = custom_code(config);

    quote! {
        #common
        #custom
    }
}

fn common_code(common: &CommonConfig) -> proc_macro2::TokenStream {
    let CommonConfig { canister_name } = common;

    quote! {
        use candid::{CandidType, Decode, Encode};
        use chainsight_cdk::{
            core::{U256},
            indexer::{Event, Indexer, IndexingConfig},
            storage::Data,
            web3::Web3CtxParam,
        };
        use chainsight_cdk_macros::{
            define_get_ethereum_address, define_transform_for_web3, define_web3_ctx, did_export, init_in,
            manage_single_state, chainsight_common, setup_func, web3_event_indexer, timer_task_func, prepare_stable_structure, stable_memory_for_scalar, web3_event_indexer_source,
            ContractEvent, Persist, StableMemoryStorable, CborSerde
        };
        use ic_solidity_bindgen::{types::EventLog};
        use ic_stable_structures::writer::Writer;
        use ic_web3_rs::{
            ethabi::Address,
            futures::{future::BoxFuture, FutureExt},
            transports::ic_http_client::CallOptions,
        };
        use serde::Serialize;
        use std::{collections::HashMap, str::FromStr};
        did_export!(#canister_name);

        // NOTE: The memory id in canister is used from a number that does not duplicate the memory id declared in the storage module of the cdk.
        // https://github.com/horizonx-tech/chainsight-sdk/blob/8aa1d1dd1cb8e3d0adde2fa9d27f374d430f663a/chainsight-cdk/src/storage/storage.rs#L97
        init_in!(11);
        chainsight_common!();
        define_web3_ctx!(12);
        define_transform_for_web3!();
        define_get_ethereum_address!();
        prepare_stable_structure!();
        stable_memory_for_scalar!("target_addr", String, 13, false);
        setup_func!({
            target_addr: String,
            web3_ctx_param: Web3CtxParam,
            config: IndexingConfig,
        }, 14);
        timer_task_func!("set_task", "index", 15);
    }
}

fn custom_code(config: EventIndexerConfig) -> proc_macro2::TokenStream {
    let EventIndexerConfig {
        common: _,
        def:
            EventIndexerEventDefinition {
                identifier,
                abi_file_path,
            },
    } = config;

    let contract_struct_name_ident =
        format_ident!("{}", extract_contract_name_from_path(&abi_file_path));

    let event = {
        let reader = File::open(Path::new(&abi_file_path))
            .unwrap_or_else(|_| panic!("Couldn't open {}", &abi_file_path));
        let binding = ethabi::Contract::load(reader)
            .unwrap_or_else(|e| panic!("Failed to load contract from abi file: {}", e));
        // NOTE: Can I use .event? https://docs.rs/ethabi/latest/ethabi/struct.Contract.html#method.event
        binding
            .events_by_name(&identifier)
            .unwrap_or_else(|_| panic!("Failed to find event by name {}", &identifier))
            .first()
            .unwrap_or_else(|| panic!("Failed to no events, name is {}", &identifier))
            .clone()
    };

    let call_func_ident = format_ident!("event_{}", camel_to_snake(&event.name));
    let (event_struct_ident, event_struct) = generate_event_struct(event);

    quote! {
        ic_solidity_bindgen::contract_abi!(#abi_file_path);
        web3_event_indexer_source!(#event_struct_ident);
        web3_event_indexer!(#event_struct_ident, 16);
        #event_struct

        fn get_logs(
            from: u64,
            to: u64,
            call_options: CallOptions,
        ) -> BoxFuture<'static, Result<HashMap<u64, Vec<EventLog>>, chainsight_cdk::indexer::Error>> {
            async move {
                let res = #contract_struct_name_ident::new(
                    Address::from_str(get_target_addr().as_str()).expect("Failed to parse target addr to Address"),
                    &web3_ctx().expect("Failed to get web3_ctx"),
                ).#call_func_ident(from, to, call_options).await;
                match res {
                    Ok(logs) => Ok(logs),
                    Err(e) => Err(chainsight_cdk::indexer::Error::OtherError(e.to_string())),
                }
            }.boxed()
        }
    }
}

fn generate_event_struct(event: ethabi::Event) -> (proc_macro2::Ident, proc_macro2::TokenStream) {
    let ethabi::Event { name, inputs, .. } = event;
    let struct_ident = format_ident!("{}", &name);
    let struct_field_tokens = inputs
        .clone()
        .into_iter()
        .map(|event_param| {
            let field_name_ident = format_ident!("{}", &event_param.name);
            let field_ty_ident = convert_event_param_type_to_field_ty_ident(&event_param.kind)
                .unwrap_or_else(|_| {
                    panic!(
                        "Failed to convert event's ParamType `{}` to field type ident",
                        &event_param.kind
                    )
                });
            quote! { pub #field_name_ident: #field_ty_ident }
        })
        .collect::<Vec<_>>();

    (
        struct_ident.clone(),
        quote! {
            #[derive(Clone, Debug,  Default, candid::CandidType, ContractEvent, Serialize, Persist)]
            pub struct #struct_ident {
                #(#struct_field_tokens),*
            }

            impl chainsight_cdk::indexer::Event<EventLog> for #struct_ident {
                fn tokenize(&self) -> chainsight_cdk::storage::Data {
                    self._tokenize()
                }

                fn untokenize(data: chainsight_cdk::storage::Data) -> Self {
                    #struct_ident::_untokenize(data)
                }
            }
        },
    )
}

fn convert_event_param_type_to_field_ty_ident(
    param_type: &ethabi::ParamType,
) -> anyhow::Result<proc_macro2::Ident> {
    let field_ty =
        convert_type_from_ethabi_param_type(param_type).map_err(|e| anyhow::anyhow!(e))?;
    let field_ty_ident = match field_ty.as_str() {
        ADDRESS_TYPE => format_ident!("String"),
        U256_TYPE => format_ident!("U256"),
        _ => format_ident!("{}", field_ty),
    };
    Ok(field_ty_ident)
}

#[cfg(test)]
mod test {
    use chainsight_cdk::config::components::{CommonConfig, EventIndexerEventDefinition};
    use insta::assert_snapshot;
    use rust_format::{Formatter, RustFmt};

    use super::*;

    #[test]
    fn test_snapshot() {
        let config = EventIndexerConfig {
            common: CommonConfig {
                canister_name: "app".to_string(),
            },
            def: EventIndexerEventDefinition {
                identifier: "Transfer".to_string(),
                abi_file_path: "examples/minimum_indexers/src/event_indexer/abi/ERC20.json"
                    .to_string(),
            },
        };
        let generated = event_indexer_canister(config);
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__event_indexer", formatted);
    }
}
