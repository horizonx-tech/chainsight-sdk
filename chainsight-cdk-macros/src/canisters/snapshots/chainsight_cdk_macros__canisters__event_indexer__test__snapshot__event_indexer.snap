---
source: chainsight-cdk-macros/src/canisters/event_indexer.rs
expression: formatted
---
use candid::CandidType;
use chainsight_cdk::{
    core::U256,
    indexer::{Event, Indexer, IndexingConfig},
    storage::Data,
    web3::Web3CtxParam,
};
use chainsight_cdk_macros::{
    chainsight_common, define_get_ethereum_address, define_transform_for_web3, define_web3_ctx,
    did_export, init_in, manage_single_state, setup_func, timer_task_func, web3_event_indexer,
    web3_event_indexer_source, ContractEvent, Persist,
};
use ic_solidity_bindgen::types::EventLog;
use ic_web3_rs::{
    ethabi::Address,
    futures::{future::BoxFuture, FutureExt},
    transports::ic_http_client::CallOptions,
};
use serde::Serialize;
use std::{collections::HashMap, str::FromStr};
did_export!("app");
chainsight_common!();
define_web3_ctx!();
define_transform_for_web3!();
define_get_ethereum_address!();
timer_task_func!("set_task", "index");
manage_single_state!("target_addr", String, false);
setup_func ! ({ target_addr : String , web3_ctx_param : Web3CtxParam , config : IndexingConfig , });
init_in!();
ic_solidity_bindgen::contract_abi!("examples/minimum_indexers/src/event_indexer/abi/ERC20.json");
web3_event_indexer_source!(Transfer);
web3_event_indexer!(Transfer);
#[derive(Clone, Debug, Default, candid :: CandidType, ContractEvent, Serialize, Persist)]
pub struct Transfer {
    pub from: String,
    pub to: String,
    pub value: U256,
}
impl chainsight_cdk::indexer::Event<EventLog> for Transfer {
    fn tokenize(&self) -> chainsight_cdk::storage::Data {
        self._tokenize()
    }
    fn untokenize(data: chainsight_cdk::storage::Data) -> Self {
        Transfer::_untokenize(data)
    }
}
fn get_logs(
    from: u64,
    to: u64,
    call_options: CallOptions,
) -> BoxFuture<'static, Result<HashMap<u64, Vec<EventLog>>, chainsight_cdk::indexer::Error>> {
    async move {
        let res = ERC20::new(
            Address::from_str(get_target_addr().as_str())
                .expect("Failed to parse target addr to Address"),
            &web3_ctx().expect("Failed to get web3_ctx"),
        )
        .event_transfer(from, to, call_options)
        .await;
        match res {
            Ok(logs) => Ok(logs),
            Err(e) => Err(chainsight_cdk::indexer::Error::OtherError(e.to_string())),
        }
    }
    .boxed()
}

